infra/ballista/build/lambda-handler.py

```diff
@@ -0,0 +1,190 @@
+import os
+import subprocess
+from subprocess import PIPE
+import logging
+import pexpect
+from pexpect import popen_spawn
+import time
+import sys
+
+
+logging.getLogger().setLevel(logging.INFO)
+
+# Create global variables for the forked processes
+process_s = None
+process_e = None
+process_cli = None
+IS_COLD_START = True
+
+
+def init():
+    # start scheduler
+    global process_s
+    process_s = subprocess.Popen(
+        ["/opt/ballista/ballista-scheduler", "--sled-dir", "/tmp/scheduler/sled"],
+        stdout=PIPE,
+        stderr=sys.stderr,
+        bufsize=0,
+    )
+    logging.info("scheduler starts")
+    time.sleep(1)
+    # wait till the scheduler is up and running
+    if process_s.poll() is None:
+        return_code = 1
+        timeout = 0
+        while return_code:
+            res = subprocess.run(
+                ["nc", "-z", "localhost", "50050"], capture_output=True
+            )
+            return_code = res.returncode
+            logging.debug(res)
+            time.sleep(1)
+            # rc = process_s.poll()
+            timeout += 1
+            if timeout >= 30:  # or rc is not None:
+                process_s.terminate()
+                logging.error(process_s.stdout.read().decode("utf-8"))
+                logging.error(process_s.stderr.read().decode("utf-8"))
+                raise Exception(f"scheduler failed to start after {timeout} seconds")
+    else:
+        logging.error(process_s.stdout.read().decode("utf-8"))
+        raise Exception(
+            f"scheduler failed to init stderror: {process_s.stderr.read().decode('utf-8')}"
+        )
+
+    global process_e
+    process_e = subprocess.Popen(
+        ["/opt/ballista/ballista-executor"],
+        stdout=PIPE,
+        stderr=sys.stderr,
+        bufsize=0,
+    )
+    logging.info("executor starts")
+    # wait till the executor is up and running
+    if process_s.poll() is None:
+        return_code = 1
+        timeout = 0
+        while return_code:
+            res = subprocess.run(
+                ["nc", "-z", "localhost", "50051"], capture_output=True
+            )
+            return_code = res.returncode
+            logging.debug(res)
+            time.sleep(1)
+            timeout += 1
+            if timeout >= 30:
+                process_e.terminate()
+                logging.error(process_e.stdout.read().decode("utf-8"))
+                logging.error(process_e.stderr.read().decode("utf-8"))
+                process_s.terminate()
+                logging.error(process_s.stdout.read().decode("utf-8"))
+                logging.error(process_s.stderr.read().decode("utf-8"))
+                raise Exception(f"executor failed to start after {timeout} seconds")
+    else:
+        logging.error(process_e.stdout.read().decode("utf-8"))
+        logging.error(process_s.stdout.read().decode("utf-8"))
+        raise Exception(
+            f"executor failed to init stderror: {process_e.stderr.read().decode('utf-8')}"
+        )
```

Comment:
❕ I find the process execution checking a bit too evolved, something simpler where you just start the process an try to connect for 15sec would be enough. 
  - For instance the wrapping `if process_s.poll() is None` is non deterministic, because it requires the process to have failed fast enough to be picked up that early, so I would argue that it is not that useful.
⁉️ the process checking for the scheduler and the executor are almost identical. They should be refactored into one common function
❓ cleaning up by terminating the process is a good idea in case you want to try to run the function again after an init failure, and it is not implemented for the other engines. I am wondering though if you wouldn't have problems with the files that were created by the killed processes, such as `/tmp/scheduler/sled`. Have you checked that?

Reply:
I'm refactoring this a little.
For the last question. yes. at each time the files created on the `sled` folder have `pseudo random names` so the only issue this might bring is on disk usage, but we're far from there. the created files are les than 1kB on the test I've performed

Reply:
> the files created on the sled folder have pseudo random names

if these are the only files created then perfect!